File: leetcode3/이진희/3880_Minimum Absolute Difference Between Two Values.java
/*

1. 아이디어 : 완전탐색으로 조건을 만족하는 최소 인덱스를 구한다

2. 시간복잡도 : O(N^2)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public int minAbsoluteDifference(int[] nums) {
        // 조건을 만족하는 최소 절댓값
        // i == 1, j == 2
        // 순서 상관 없음

        return findMinDiff(nums);
    }

    private int findMinDiff(int[] nums) {
        // 0, 1, 2만 존재
        int ans = 1000;
        for(int i=0; i<nums.length-1; i++) {
            for(int j=i+1; j<nums.length; j++) {
                if((nums[i] == 1 && nums[j] == 2) || (nums[i] == 2 && nums[j] == 1)) {
                    ans = Math.min(ans, Math.abs(i-j));
                }
            }
        }

        if(ans == 1000) return -1;
        else return ans;
    }
}

File: leetcode3/최원준/3880. Minimum Absolute Difference Between Two Values.py
#

'''
1. 아이디어 :
nums를 순회하면서 1과 2를 마주칠때마다 
- 1 또는 2의 마지막 인덱스와의 거리를 계산합니다.
- 인덱스를 저장합니다.


2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :


'''
class Solution:
    def minAbsoluteDifference(self, nums: list[int]) -> int:
        index1 = float('inf')
        index2 = float('inf')
        ans = float('inf')
        
        n = len(nums)
        for i in range(n):
            num = nums[i]
            if num == 1:
                ans = min(ans, abs(i-index2))
                index1 = i
            elif num == 2:
                ans = min(ans, abs(i-index1))
                index2 = i
        return -1 if ans == float('inf') else ans
